### PR TITLE
limit cross dc gossip johanandren

### PR DESCRIPTION
**Pull request by johanandren**
Refs #23282

### CODE REVIEW

File: akka-cluster/src/main/resources/reference.conf
Comment (patriknw) [inline akka-cluster/src/main/resources/reference.conf]: split this into two sentences, second "Used for gossip and heartbeating."

**Comment by johanandren** [comment]
Current thing: if node is among N oldest in DC it will create a set of all the other N oldest + the local nodes and randomly gossip to one of those if it is not one of the N oldest it just gossips among the local nodes.

**Comment by johanandren** [comment]
Depends on #23288 being fixed before it will work though.

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: It's a best effort when shutting down. It could be good to do some of both, e.g. when stopping last node in a dc it can be good that it's spread to other dc.
1 to other dc, and the rest to own dc?

File: akka-cluster/src/main/resources/reference.conf
Comment (ktoso) [inline akka-cluster/src/main/resources/reference.conf]: me likes, good names

File: akka-cluster/src/main/resources/reference.conf
Comment (ktoso) [inline akka-cluster/src/main/resources/reference.conf]: I'd rather say "If the cluster does not span multiple data centers, this value has no effect"?

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: isn't the difference between `gossipTo` and `gossipStatusTo` lost?

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: `!nodesPerDc(selfTeam).contains(selfMember)`

but we don't have a `selfMember`, I might introduce that when I do the `Gossip` refactoring

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: I think we should gossip less often to other DCs, so the round could start with a decision 80% local or 20% cross.

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala
Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: Yes, I missed out on that one, damn.

File: akka-cluster/src/main/scala/akka/cluster/ClusterSettings.scala
Comment (ktoso) [inline akka-cluster/src/main/scala/akka/cluster/ClusterSettings.scala]: 👍

**Review by ktoso — APPROVED** [review]
LGTM, looks good

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: a slightly other way would be, for each gossip round:

* decide own DC or other DC
  * if own => as before
  * if other => select one of the other DCs with equal probably
    * find the CrossDcConnections oldest for that DC, and then pick one random of them

I'm not sure what is best. If some DC contains many more nodes than another this approach might be better for spreading updates to all DCs quickly?

Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: I was thinking about something like that but then I thought about the case where all those nodes are unavailable and having to "try again" with another DC which the current scheme avoids.  With the "many more nodes" it is still limited by the max-connection setting so as long as we choose a fairly low number for that it wouldn't make a big difference. 

I agree on the 80/20 though. Will add that.

Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: Thought about it some more and now I agree that choosing a DC randomly first is better because of not preferring DCs with more nodes than the limit over DCs with fewer nodes. I'll update the logic.

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala
Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: Is there a reason to not always do this when we have a potential node (if the new localDcGossipTarget already prefers nodes that hasn't seen the gossip?)

Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: not sure what you mean, but we shouldn't change how the gossip interaction works within same DC, for the cross DC I think we can always use `gossipStatusTo` and don't care about the `seen` stuff.

Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: In the current logic we mix selection and side effect, which I was trying to get away from. I moved this all-nodes-has-seen-gossip fallback inside of the logic that gives back the set of nodes to choose randomly among. In that case the side-effect logic doing the actual gossip doesn't know about which list of nodes that was chosen, so I wondered if after selecting that random node, it would hurt to too the seenByNode check and if seen only gossip status. I think it should give the same logic as currently for local (and do that same thing for cross-dc gossip)

Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: Ultimately I would want those two selection methods pure and then cover with unit tests

Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: Doing the `seenByNode` twice shouldn't be a problem, it's just a Map lookup.

For the cross DC it's much more likely that the target anyway has the latest state even though it's not marked as such in the seen yet, and therefore I think it makes sense to always use the `gossipStatusTo` to avoid transferring the full thing unless necessary.

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: `else firstSelection` :)

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: add config, we have similar `gossip-different-view-probability`

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: tailrec

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: as mentioned elsewhere I don't think we should care about seen for cross dc

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: You do a lot of stuff with the sorting and finding nodes in other DCs than the one you finally select.
You can select DC first, and then only find X oldest for that DC.

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: clean 👍

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: shouldn't be a big deal

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: ok, if we cache them in MembershipState this is fine

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: seen for cross dc?

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: `ThreadLocalRandom.current.nextInt(nodes.size)`
(yeah, I know it was not written by you)

**Review by patriknw — COMMENTED** [review]
looking good, should only be some small polish left

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: I concur! :)

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: Ah, we ignore seen for cross dc and always send GossipStatus, yes... forgot about that.

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: I realised something here: I was taking up to the max nr of interconnections when creating the map. But I think it should really be all nodes per DC sorted by age in the map since filtering is done on whether the nodes are suitable for gossip, and the number limit is among the first N "gossipable" nodes, WDYT?

Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: no, we don't want to increase the number of connections, e.g. if two are temporarily unreachable we don't want to start 2 new connections, and in the end have 7

there is a tricky case that if all 5 are stopped at the same time the other side will not know. I don't see a good solution for that, apart from documenting that oldest should be kept until last for rolling upgrades (and don't stop all at the same time)

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: `isReachableExcludingDownedObservers` will not cover cross DC nodes, for them I think we should use `isReachable(selfUniqueAddress, node)`

Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: Ah, so actual pair wise reachability observations then, hmm. Well as failure detection should happen from the same N node I guess if link is down there will be exactly those pairs?

That invalidates my previous thought about limiting to the N first gossipable nodes though.

Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: > failure detection should happen from the same N

right, that is my thinking

> limiting to the N first gossipable nodes 

The problem with that is that when there is a real network partition we will eventually try with all.

Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: I pushed something that specifically broke that last night, I'll undo that/change it back to the old (right) way of selecting nodes then.

File: akka-cluster/src/main/resources/reference.conf
Comment (ktoso) [inline akka-cluster/src/main/resources/reference.conf]: I think we'll move that into a section like `multi-data-center` in my PR i have that section but let's fix that later btw

Comment (johanandren) [inline akka-cluster/src/main/resources/reference.conf]: sure

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: update comment

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: config for 0.2, there are some similar config for the other probabilities

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]:
```
if (selectedNodes.size <= n) selectedNodes
else Random.shuffle(selectedNodes).take(n)
```

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: take(n - 1) ?

File: akka-cluster/src/multi-jvm/scala/akka/cluster/MultiDcClusterSpec.scala
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/MultiDcClusterSpec.scala]: but when we have the right think implemented this should only include the unreachable within the same dc, so this check will be rather fruitless

File: akka-cluster/src/main/resources/reference.conf
Comment (ktoso) [inline akka-cluster/src/main/resources/reference.conf]: Good, I've also set it to 5 for the heartbeating.
What about logging a warning if this number is lower than the heartbeating nodes number?

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (ktoso) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: N-th

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (ktoso) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: perhaps refactor the method a bit so you always select the local ones (it's the same code and not super trivial line of code, and then only add the other-dc `if (probablility check)`

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (ktoso) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: this ~~will~~ means

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala
Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: that's what I thought I did write, yes :)

File: akka-cluster/src/multi-jvm/scala/akka/cluster/MultiDcClusterSpec.scala
Comment (johanandren) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/MultiDcClusterSpec.scala]: Yeah, this is more as a guard to know when to move on, I guess we will have to update these tests when that is in place.

File: akka-cluster/src/main/resources/reference.conf
Comment (johanandren) [inline akka-cluster/src/main/resources/reference.conf]: I think we should use the same setting or have it derived, because the whole idea is to gossip and failure detect the same nodes.

**Review by ktoso — COMMENTED** [review]
Some silly comments, feel free to address whenever

File: akka-cluster/src/main/resources/reference.conf
Comment (ktoso) [inline akka-cluster/src/main/resources/reference.conf]: +1 for using the same setting 👍

**Review by ktoso — APPROVED** [review]

**Comment by johanandren** [comment]
All cluster tests green locally as well.

**Comment by johanandren** [comment]
After rebasing on top of merged heartbeating PR the following multi node tests fail:
 * akka.cluster.MultiDc
 * akka.cluster.MultiDcSplitBrain
 * akka.cluster.MultiDcFewCrossDc
 * akka.cluster.MultiDcSunnyWeather

**Comment by patriknw** [comment]
It could be because we haven't merged Johannes' PR yet and then the unreachability observations across teams will become ordinary within team unreachability. I think you can merge his PR, and we can follow up with more tests in another PR.

**Comment by johanandren** [comment]
It was the unreachability assertions you mentioned earlier, at least a few of the tests, in the process of fixing them now. Will merge after that and if there is a test failing for other reasons I will investigate after merging.

**Comment by johanandren** [comment]
`akka.cluster.MultiDcSunnyWeather` still failing now, will look at that separately after merging

**Pull request merged**
